src/1-theory/3-input-and-output.adoc

```diff
@@ -0,0 +1,593 @@
+These days it's far more common to utilize HTTP properly, and give the
+response a status code as defined in the RFC have a number from `200` to
+`599` -- with plenty of gaps in between -- and each has a message and a
+definition. Most server-side languages, frameworks, etc., default to
+`200 OK`.
+
+Status codes are grouped into a few different categories:, with the
+first number being an identifier of the category of thing that happened.
+
+=== 2XX is all about success
+
+Whatever your application tried to do was successful, up to the point that the
+response was sent. A `200 OK` means you got your answer, a `201 Created` means
+the thing was created, and a `202 Accepted` is similar but does not say anything
+about the actual result, it only indicates that a request was accepted and is
+being processed asynchronously. It could still go wrong, but at the time of
+responding it was all looking good so far.
+
+=== 3XX is all about redirection
+
+These are all about sending the calling application somewhere else for the
+actual resource. The best known of these are the `303 See Other` and the `301
+Moved Permanently`, which are used a lot on the web to redirect a browser to
+another URL. Usually a redirect will be combined with a `Location` header to
+point to the new location of the content.
+
+=== 4XX is all about client errors
+
+Indicate to your clients that they did something wrong. They might have
+forgotten to send authentication details, provided invalid data, requested a
+resource that no longer exists, or done something else wrong which needs fixing.
+
+=== 5XX is all about service errors
+
+With these status codes, the API, or some network component like a load
+balancer, web server, application server, etc. is indicating that something went
+wrong in their side. For example, a database connection failed, or another
+service was down. Typically, a client application can retry the request. The
+server can even specify when the client should retry, using a `Retry-After` HTTP
+header.
+
+=== Common Status Codes
+
+Arguments between developers will continue for the rest of time over the
+exact appropriate code to use in any given situation, but these are the
+most important status codes to look out for in an API:
+
+* 200 - Generic everything is OK
+* 201 - Created something OK
+* 202 - Accepted but is being processed async (for a video means
+encoding, for an image means resizing, etc.)
+* 400 - Bad Request (should really be for invalid syntax, but some folks
+use for validation)
+* 401 - Unauthorized (no current user and there should be)
+* 403 - The current user is forbidden from accessing this data
+* 404 - That URL is not a valid route, or the item resource does not
+exist
+* 405 - Method Not Allowed (your framework will probably do this for
+you)
+* 409 - Conflict (Maybe somebody else just changed some of this data, or
+status cannot change from e.g: "published" to "draft")
+* 410 - Data has been deleted, deactivated, suspended, etc.
```

Comment:
I think it would be nice to include the official iana Description in all of these, so people can associate the formal name with the explanation in their heads more easily. Or it might just be me. :P Either way something like 
```suggestion
* 410 - Gone - Data has been deleted, deactivated, suspended, etc.
```